### PR TITLE
feat(api): add nvim__redraw for more granular redrawing

### LINKED ISSUE #20793: cursor doesn't redraw with getchar
### Neovim version (nvim -v)

nightly

### Vim (not Nvim) behaves the same?

not relevant

### Operating system/version

linux

### Terminal name/version

kitty

### $TERM environment variable

xterm-kitty

### Installation

bob

### How to reproduce the issue

`:luafile %` the file below

```lua
vim.fn.getcharstr()
vim.api.nvim_win_set_cursor(0, { 1, 10 })
-- after this, the cursor will not be shown at {1,10}
vim.fn.getcharstr()
-- now it will show at the correct position
print("done")
```

### Expected behavior

After the first `getcharstr`, the visible cursor should be at `{1,10}`

### Actual behavior

The cursor is still shown at the original position and will move to `{1,10}` after the second `getcharstr` finishes.

### PR DESCRIPTION
**Pull request by luukvbaal**
Experimental and subject to future changes.
Add a way to redraw certain elements that are not redrawn while Nvim is waiting
for input, or currently have no API to do so. This API covers all that can be
done with the :redraw* commands, in addition to the following new features:
- Immediately move the cursor to a (non-current) window.
- Target a specific window or buffer to mark for redraw.
- Mark a buffer range for redraw (replaces nvim__buf_redraw_range()).
- Redraw the 'statuscolumn'.

Resolve #20793
Closes #26071

### CODE REVIEW

File: runtime/doc/api.txt
Comment (clason) [inline runtime/doc/api.txt:1315]:
```suggestion
                • `screen`: Immediately update the screen. Useful when
```
update (=redraw) is implied

Comment (clason) [reply runtime/doc/api.txt:1315]: In general, we should try to be consistent about "update" vs. "redraw" here (or, if there is a distinction, explain it).

Comment (luukvbaal) [reply runtime/doc/api.txt:1315]: The implication of `update_screen` is that the actual buffer content in windows is redrawn, for which `update_screen()` is the only entry point internally. As opposed the various other "widgets" associated with a window which can also be targeted in this API and redrawn without redrawing the entire window, but would also be redrawn when `update_screen()` is called and they were pending for redraw.

Considering that, maybe this can be called `text` or `viewport` as it is referred to under `:h window`?

Comment (clason) [reply runtime/doc/api.txt]: Thanks for the explanation. So `flush` is (to me!) closest in meaning (but still needs an explanation at this point in the documentation, or at least a taglink), but I defer to @bfredl

File: src/nvim/api/vim.c
Comment (bfredl) [inline src/nvim/api/vim.c:2464]: I think it should be enough to do these around `update_screen()` inside its if-statement.

File: src/nvim/api/vim.c
Comment (bfredl) [inline src/nvim/api/vim.c:2484]: if we treat `win` and `buf` as mutually exclusive, we should throw an error when both `buf` and `win` are supplied.

File: src/nvim/api/vim.c
Comment (bfredl) [inline src/nvim/api/vim.c:2499]: then, if `win` is supplied but not `buf`, this either needs to throw an error or use the buffer of that window (ideally it would imply only redrawing the range for the specific window, but doing it for the buffer in the window is good enough).

File: src/nvim/api/vim.c
Comment (luukvbaal) [reply src/nvim/api/vim.c:2464]: `win_redr_status()` checks `!redrawing()` as well.

File: src/nvim/api/vim.c
Comment (luukvbaal) [reply src/nvim/api/vim.c:2499]: Yes, I considered exposing `win_upate()` to be able to target specific windows but I'm not sure that's a good idea. It would only be effective when `redrawing() == false` anyways, otherwise the rest of the windows would be updated at the next `update_screen()` anyways.

File: runtime/doc/api.txt
Comment (zeertzjq) [inline runtime/doc/api.txt:1329]: I don't think adding these new abbreviations is a good idea. I think it's better to use existing terminologies.

Comment (zeertzjq) [reply runtime/doc/api.txt:1329]: I guess it's better to put them directly in `opts`, so that there will be fewer thing to decide when something new need to be added.

File: src/nvim/api/vim.c
Comment (luukvbaal) [reply src/nvim/api/vim.c:2518]: No it's intentionally `!=`. It should still redraw the others if they are supplied.

File: src/nvim/api/vim.c
Comment (zeertzjq) [inline src/nvim/api/vim.c:2453]: Maybe it's better to just mention here that these two are used a target a specific window/buffer for other optional parameters (without mentioning their names), and make it clear in each optional parameter how it uses `buf`/`win`.

File: src/nvim/api/vim.c
Comment (zeertzjq) [inline src/nvim/api/vim.c:2442]: There seems to be some line wrapping problems with docgen, plus there is no need to mention "line number of the range" twice
```suggestion
///                 providers. Tuple `[first, last]` are the first and last
///                 line numbers of the range, respectively.
```

File: src/nvim/api/vim.c
Comment (zeertzjq) [inline src/nvim/api/vim.c:2493]:
```suggestion
             "%s", "Invalid 'range': expected 2-tuple of Integers", {
```

Comment (luukvbaal) [reply src/nvim/api/vim.c:2493]: Seemed more logical to me as well but then this should be updated first I think:
https://github.com/neovim/neovim/blob/8697f3274b240a5a00774dba90329c00c4e914ab/src/nvim/lua/converter.c#L829

File: src/nvim/api/vim.c
Comment (zeertzjq) [inline src/nvim/api/vim.c:2470]: API error messages usually don't end with a period
```suggestion
    VALIDATE(win == NULL, "%s", "Invalid 'buf': already specified 'win'", {
```

**Comment by zeertzjq** [comment]
It seems that for some reason the Perl tests become very flaky on this PR. I guess that may have something to do with the added API metadata.

**Comment by luukvbaal** [comment]
Is that our problem or should the provider handle it?

**Comment by zeertzjq** [comment]
Not sure, but anyway it's out of the scope of this PR.

File: src/nvim/api/deprecated.c
Comment (justinmk) [inline src/nvim/api/deprecated.c:704]: this is a private api, can we just remove it now?

**Review by justinmk — APPROVED** [review]

File: src/nvim/api/deprecated.c
Comment (luukvbaal) [reply src/nvim/api/deprecated.c:704]: Fine by me but considering that it is used in the wild a single release cycle with `@deprecated` could be good?

Comment (justinmk) [reply src/nvim/api/deprecated.c:704]: In general we should be pretty strict about not making efforts (deprecation cycles) for APIs that were marked as private. I see some important plugins in there, though I think it will be easy for them to migrate...

File: src/nvim/api/vim.c
Comment (luukvbaal) [inline src/nvim/api/vim.c:2487]: I think it should be reconsidered what `first` and `last` actually indicate, and how they are documented. It seems that in `nvim__buf_redraw_range` `first` was a 0-indexed row number, whereas `last` was a 1-indexed line number (why?). I'll go ahead and document it and update the internal calls like so:

```
- range: Redraw a range in `buf`, the buffer in `win` or
  the current buffer. Useful for decoration providers.
  Expects a tuple `[first, last]` with the first and last
  line number of the range, or [1, -1] to indicate the entire
  buffer.
```

Comment (luukvbaal) [reply src/nvim/api/vim.c]: Ah yeah that makes more sense probably, so then they were both 0-indexed. But I think 1-indexed end-inclusive line numbers makes more sense for this API?

Comment (luukvbaal) [reply src/nvim/api/vim.c]: I mean in a way it doesn't because the extmarks that necessitate this call also were placed 0-indexed (end inclusive). I just mean in general when talking about line numbers 1-indexed feels more natural but if this should be 0-indexed as well it just needs to be documented.

Comment (luukvbaal) [reply src/nvim/api/vim.c]: It's also not entirely clear to me why this API is needed in the first place so that probably doesn't help. Don't the `nvim_buf_set_extmark()` calls already set `w_redraw_top/mod` through `decor_redraw()`? If this API is graduated the reasoning behind that should also probably be documented better (here)?

Comment (luukvbaal) [reply src/nvim/api/vim.c]: I think all internal `nvim__buf_redraw_range(buf, 0, -1)` calls could/should be replaced by something like:

```diff
@@ -2425,8 +2425,9 @@ static void redraw_status(win_T *wp, Dict(redraw) *opts, bool *flush)
+///               - valid: When present and `true`, redraw the changed part of `win`,
+///                 all windows containing `buf`, or all windows. When `false`, redraw
+///                 the entire window.
 ///               - range: Redraw a range in `buf`, the buffer in `win` or the
 ///                 current buffer. Useful for decoration providers. Expects a
 ///                 tuple `[first, last]` with the first and last line number of
@@ -2462,13 +2463,14 @@ void nvim_redraw(Dict(redraw) *opts, Error *err)
     }
   }

-  if (opts->clear) {
+  if (HAS_KEY(opts, redraw, valid)) {
+    int type = opts->valid ? UPD_VALID : UPD_NOT_VALID;
     if (win != NULL) {
-      redraw_later(win, UPD_NOT_VALID);
+      redraw_later(win, type);
     } else if (buf != NULL) {
-      redraw_buf_later(buf, UPD_NOT_VALID);
+      redraw_buf_later(buf, type);
     } else {
-      redraw_all_later(UPD_NOT_VALID);
+      redraw_all_later(type);
     }
   }
```

Then, if placing ephemeral marks would call `decor_redraw()`, I don't think this `range` API would be needed.

Comment (luukvbaal) [reply src/nvim/api/vim.c]: It also seems like the `nvim__buf_redraw_range(buf, 0, -1)` calls were "abusing" the private API. I will try to replace `clear` with `valid` in a way that I think should cover that use case.

File: src/nvim/api/vim.c
Comment (justinmk) [inline src/nvim/api/vim.c:2423]:
```suggestion
///
/// @see |:redraw|
///
```

Let's also mention `nvim_redraw()` from the docs for :redraw/:redrawstatus/etc.

Comment (luukvbaal) [reply src/nvim/api/vim.c:2423]: Sure. Could add a line like this to each command

    See |nvim_redraw()| for a more granular alternative.

Or add some kind of header above the bunch of them as those commands are grouped.

Comment (justinmk) [reply src/nvim/api/vim.c:2423]: > Could add a line like this to each command

Just `See also |nvim_redraw()|.` on each command is the usual pattern. Headers are less effective for "discovery".

File: runtime/doc/api.txt
Comment (justinmk) [inline runtime/doc/api.txt:1311]: If {opts} is empty what does it do?
```suggestion
    Instruct Nvim to redraw various components. By default (i.e. if {opts} is empty), ...
```

File: src/nvim/api/deprecated.c
Comment (justinmk) [inline src/nvim/api/deprecated.c:16]: is this still needed?

File: test/functional/api/vim_spec.lua
Comment (justinmk) [inline test/functional/api/vim_spec.lua:4999]: is there a test for empty `opts` ? `nvim_redraw({})`

**Comment by justinmk** [comment]
In order to unblock this and get it in the tree, shall we rename it to double-underscore `nvim__redraw`? That signals that it is experimental. Then we can finalize it in 0.11

**Comment by luukvbaal** [comment]
Fine by me, updated. Not sure whether to keep the `various.txt` references. Referring to experimental API is a little weird but removing them risks forgetting them in the future.

**Comment by clason** [comment]
Yeah, I'd avoid the links from `various.txt` for now.

**Comment by justinmk** [comment]
It's marked as experimental, I don't see why we can't refer to it from other parts of the docs. If it were deprecated, that would be different, then we want to scrub the docs.

**Comment by pidgeon777** [comment]
In my setup `nvim --version`:

```log
NVIM v0.10.0-dev-3057+g9e2f378b6
Build type: RelWithDebInfo
LuaJIT 2.1.1713484068
Run "nvim -V1 -v" for more info
```

and now it seems that `Neovim` is alot faster (writing in insert mode, copy pasting in normal mode and buffer switching/navigation). Could it be due to this merge?

I'm using Windows 11 Pro.

**Comment by luukvbaal** [comment]
Uhm since this flushes the UI in addition to marking a buffer range for redraw, and it is used in the treesitter highlighter, an (unintended but wanted?) consequence could be that it made the editing experience feel more responsive (at the cost of more CPU cycles).

**Comment by pidgeon777** [comment]
I have to say that it seems to have improved `Neovim` smoothness a lot, at least on Windows (I'm also using [Nvy](https://github.com/RMichelsen/Nvy) GUI).

Or, it could also be that this merge in conjunction with other recent ones (to master) improved the overall performance (I'm using the `Neovim` nightly version currently available).

File: src/nvim/api/vim.c
Comment (zeertzjq) [inline src/nvim/api/vim.c:2419]: This should use the same buffer as the buffer passed to the `redraw_buf_range_later()` below. And I don't think it should treat all negative value as the same.

Comment (luukvbaal) [reply src/nvim/api/vim.c:2419]: Good catch. Only -1 is a valid negative value, do you just mean that the if statement should be more explicit in that?

Comment (zeertzjq) [reply src/nvim/api/vim.c:2419]: Oops, I didn't see the `>= -1` check above. Then I think just changing this < 0` to `== -1` is enough.

**Comment by topaxi** [comment]
Nightly started crashing for me when using a mouse click into neovim. After bisecting from yesterday to today, I found this PR/commit 7b14eb543d43344e2498335dc93a68d200469516 to be the cause.

I just get a segmentation fault and am not versed enough in C to be able to debug beyond that.

If I can provide anything that might help, let me know. I use quite a few plugins, and with a `--clean` run, the crash does not occur, so I assume this might be either a configuration issue or combined with a plugin.

I wasn't able to get anything out of any log.

**Comment by zeertzjq** [comment]
It may be https://github.com/folke/flash.nvim/issues/333

**Comment by topaxi** [comment]
Disabled `flash.nvim` and the crash still occurs, I might find time later today to enable/disable plugins one-by-one to find the offending one :)

**Comment by zeertzjq** [comment]
Another possible cause is https://github.com/folke/noice.nvim/blob/0cbe3f88d038320bdbda3c4c5c95f43a13c3aa12/lua/noice/util/ffi.lua#L20

**Comment by topaxi** [comment]
Good catch, your quick! That's the one!

**Comment by clason** [comment]
PSA: If things crash, grep your plugin folder for `ffi`. If a plugin uses that (and many of Folke's do), then this is likely the fault and you should report it there. We have zero ABI compatibility guarantees, and _any_ commit may break FFI. Plugin authors should know that and react quickly.

**Comment by justinmk** [comment]
coverity:

```
*** CID 493189:    (FORWARD_NULL)
/src/nvim/api/vim.c: 2419 in nvim__redraw()
2413                  "%s", "Invalid 'range': Expected 2-tuple of Integers", {
2414           return;
2415         });
2416         linenr_T first = (linenr_T)kv_A(opts->range, 0).data.integer + 1;
2417         linenr_T last = (linenr_T)kv_A(opts->range, 1).data.integer;
2418         if (last < 0) {
>>>     CID 493189:    (FORWARD_NULL)
>>>     Dereferencing null pointer "buf".
2419           last = buf->b_ml.ml_line_count;
2420         }
2421         redraw_buf_range_later(win ? win->w_buffer : (buf ? buf : curbuf), first, last);
2422       }
2423     
2424       if (opts->cursor) {
/src/nvim/api/vim.c: 2419 in nvim__redraw()
2413                  "%s", "Invalid 'range': Expected 2-tuple of Integers", {
2414           return;
2415         });
2416         linenr_T first = (linenr_T)kv_A(opts->range, 0).data.integer + 1;
2417         linenr_T last = (linenr_T)kv_A(opts->range, 1).data.integer;
2418         if (last < 0) {
>>>     CID 493189:    (FORWARD_NULL)
>>>     Dereferencing null pointer "buf".
2419           last = buf->b_ml.ml_line_count;
2420         }
2421         redraw_buf_range_later(win ? win->w_buffer : (buf ? buf : curbuf), first, last);
2422       }
2423     
2424       if (opts->cursor) {
```

**Comment by luukvbaal** [comment]
Was fixed in https://github.com/neovim/neovim/pull/28614 @justinmk.

**Pull request merged**
